marlowe-playground-client/src/Projects/View.purs

```diff
@@ -0,0 +1,104 @@
+import Data.Formatter.DateTime (FormatterCommand(..), format)
+import Data.Lens (to, view, (^.))
+import Data.List (fromFoldable)
+import Data.Maybe (Maybe(..))
+import Data.Newtype (unwrap)
+import Effect.Aff.Class (class MonadAff)
+import Gist (Gist, gistCreatedAt, gistDescription, gistId, gistUpdatedAt)
+import Halogen (ClassName(..), ComponentHTML)
+import Halogen.Classes (fontSemibold, modalContent, paddingRight, smallPaddingRight, textSm)
+import Halogen.HTML (HTML, a, a_, div, div_, span, text)
+import Halogen.HTML.Events (onClick)
+import Halogen.HTML.Properties (class_, classes)
+import MainFrame.Types (ChildSlots)
+import Marlowe.Gists (fileExists, filenames, playgroundGist)
+import Modal.ViewHelpers (modalHeaderWithClose)
+import Network.RemoteData (RemoteData(..))
+import Prim.TypeError (class Warn, Text)
+import Projects.Types (Action(..), Lang(..), State, _projects, modalIsLoading)
+import Text.Parsing.Parser (runParser)
+
+render ::
+  forall m.
+  Warn (Text "Fix 'body Loading' design") =>
+  MonadAff m =>
+  State ->
+  ComponentHTML Action ChildSlots m
+render state =
+  if modalIsLoading state then
+    text mempty
+  else
+    div_
+      [ modalHeaderWithClose "Open Project" Cancel
+      , div [ classes [ modalContent ] ]
+          [ body playgroundGists
+          ]
+      ]
+  where
+  playgroundGists = filter playgroundGist <$> state ^. _projects
+
+  body (Success []) = span [ class_ (ClassName "empty-result") ] [ text "No saved projects found" ]
+
+  body (Success gists) = projectList gists
+
+  body (Failure _) = span [ class_ (ClassName "error") ] [ text "Failed to load gists" ]
+
+  body _ = text mempty
+
+projectList ::
+  forall p.
+  Warn (Text "Only 30 projects seem to be loading. Check why and define what should we do about it") =>
+  Array Gist ->
+  HTML p Action
+projectList gists =
+  div [ classes [ ClassName "project-list" ] ]
+    (headers <> projects)
+  where
+  headers :: Array (HTML p Action)
+  headers =
+    [ "Name", "Created", "Last Updated", "Open" ]
+      <#> \name ->
+          div [ classes [ textSm, fontSemibold ] ] [ text name ]
+
+  projects :: Array (HTML p Action)
+  projects =
+    gists
+      >>= \gist ->
+          [ div [ classes [ ClassName "project-name", paddingRight ] ] [ gist ^. (gistDescription <<< to text) ]
+          , div [ classes [ ClassName "date", paddingRight ] ] [ gist ^. (gistCreatedAt <<< to formatDate <<< to text) ]
+          , div [ classes [ ClassName "date", paddingRight ] ] [ gist ^. (gistUpdatedAt <<< to formatDate <<< to text) ]
+          , loadLink gist Javascript [ smallPaddingRight, fontSemibold ] $ fileExists filenames.javascript gist
+          , loadLink gist Haskell [ smallPaddingRight, fontSemibold ] $ fileExists filenames.haskell gist
+          , loadLink gist Marlowe [ smallPaddingRight, fontSemibold ] $ fileExists filenames.marlowe gist
+          , loadLink gist Blockly [ fontSemibold ] $ fileExists filenames.blockly gist
```

Comment:
I disagree, notice that the first three columns have a medium padding, while the first three actions have smallPadding and the last action has no padding. Writting that in CSS is very hard. Specially with CSS grid, in which all elements are direct children of the grid component, which makes it easier to re-arrange for different media queries

Reply:
It feels to me like a lot of the CSS community is trying to separate markup and layout/styling. This approach seems like a good idea to me, rather than having lots of rules obfuscating the markup the layout is defined separately and the style is also defined separately. These classes are the opposite approach and it just feels "wrong" to me, although I don't know enough about CSS to say really.

Reply:
I'm not a CSS expert myself so take whatever I say with a pinch of salt. 

Granted, since the early stages of web development we were taught to clearly separate the markup from the styling and logic, and websites like [css zen garden](http://www.csszengarden.com/) made a good point about how the same markup with different styles could provide completely different experiences. But in practice, I've never seen a need for that, at most having different layouts strategies for responsive sites and different colours for themes/dark-mode (which are supported by tailwind). 

IMO, using the traditional approach exposes two big problems.

## Naming is hard
You start easy with stuff like `modal`, `badge`, but end up with classes like `modal-top-container` or similar just because you need to reference a particular `div`. The issue is better described in [this documentation page](https://tailwindcss.com/docs/utility-first), but even in this thread we had a good example:

In this grid I need the first 3 columns to have "big padding", the next 3 columns to have "small padding" and the last element to have no padding at all. I can't use `:nth-child` directly because all the elements in the grid need to be direct children, (which is better for responsive design) and `:nth-child` does not support `mod` operations. So if I wanted to do this in a traditional way, what name should I use? And would that name describe the design/layout intent? The utilities are verbose but very clear in intent.

## Cascading should probably be an anti-pattern
This is a very personal opinion, but I think the cascading part of `CSS` was thougt at the same time that Java and OOP were starting to get momentum, and I think they both suffer from the same issue: "You wanted a banana but what you got was a gorilla holding the banana and the entire jungle".

I think this problem is one of the biggest reasons people dislike doing CSS, and it's best described by this Peter Griffin gif. When you modify attributes in one class expecting to change how a single component looks, you end up breaking the styles in a component which may be in a different page.

![peter](https://user-images.githubusercontent.com/2634059/103032301-7e0cf780-453e-11eb-9b4a-3746b610ffee.gif)


I attribute this problem to style inheritance, which is very dependant on the markup context (where does this component live, who are their parents). This is normally solved by overriding properties using more specific rules, but this causes specificity problems that may end up with some `!important` hacks.

To cope with this, different frameworks have different approaches:
* The [BEM](http://getbem.com/introduction/) pattern provide us with rules to generate names like `.button--state-success`. It "solves" the problem of naming being hard, but IMO, this is a recipe to provide consistent ugly names.
* In react there are different approaches for doing "CSS in JS", like [styled components](https://styled-components.com/) which end up creating unique classnames like `sc-fVOeaW`.
* [Svelte](https://svelte.dev/) and [Vue](https://vue-loader.vuejs.org/guide/scoped-css.html#scoped-css) have component-scoped CSS as a native part of the framework.

I admit that the `utility/functional css/atomic css` approach may look weird and antinatural at first, but as the author of Tailwind recently replied in [this tweet](https://twitter.com/adamwathan/status/1341144649319260168), "it's ugly but it's worth it".

One of the benefits I like the most about this approach is that "it restricts" what you should use for styling. It's like having "CSS the good parts"

![javascript_the_good_parts](https://user-images.githubusercontent.com/2634059/103032685-74d05a80-453f-11eb-8113-ff262e942ce9.jpg)


For example, the [responsive variants](https://tailwindcss.com/docs/responsive-design#mobile-first) forces you to take a mobile-first approach, which is good as it's easier to go from mobile to desktop than desktop to mobile. The [inset](https://tailwindcss.com/docs/top-right-bottom-left#usage) utilities give you sensible defaults to use to positioning absolute elements in relative parents. The [font](https://tailwindcss.com/docs/font-size), [color](https://tailwindcss.com/docs/customizing-colors) and [spacing](https://tailwindcss.com/docs/customizing-spacing) modules give you consistent scales that make the site look more homogeneous. 

Another cool benefit is that it provides constant css size. As this [post presents](https://medium.com/@johnpolacek/by-the-numbers-a-year-and-half-with-atomic-css-39d75b1263b4), traditional css tends to get larger when more features are implemented, but in utility-based frameworks the CSS size is constant, and we could configure tools to [purge](https://tailwindcss.com/docs/optimizing-for-production#writing-purgeable-html) the unused utilities.